[AndroidCrypto] Implement Root (read-only) and CurrentUser/My certificate stores

elinor-fung opened the pull request:
Implement X509 stores on Android:
- `Root`: use the AndroidCAStore (Android's store of trusted root certs)
  - Both `CurrentUser` and `LocalMachine` are read-only. Android blocks the ability to change these except through the settings application that comes with the OS. 
- `Disallowed`: match Unix behaviour - read-only, empty store for `CurrentUser`, exception for `LocalMachine`
- `CurrentUser\My`: use the AndroidKeyStore as the default store for the current application
- Add tests

Custom stores are not implemented.

cc @jkoritzinsky @steveisok @AaronRobinsonMSFT @bartonjs

elinor-fung committed:
[AndroidCrypto] Implement read-only trusted root certificate store

elinor-fung commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.cs:
```diff
@@ -44,7 +44,47 @@ public static IExportPal LinkFromCertificateCollection(X509Certificate2Collectio
+            {
+                case StoreLocation.CurrentUser:
+                {
+                    // Matches Unix behaviour of getting a disallowed store that is always empty.
```
I haven't dealt with the `My` store in this change.

From what I can tell, when Android was using OpenSSL directly, the `My` store for `CurrentUser` was really for the current application (started out empty, not shared with other applications, removed on app uninstall/reinstall). I can match that behaviour and `My`+`CurrentUser` on Android will really mean current application?

bartonjs replied:
CurrentUser\My should be wherever a normal Java application would put/get certificates it uses.  That... might... be the "PKCS12" KeyStore?  Or the "give me the default KeyStore" KeyStore?  Whatever is "natural".

bartonjs replied:
e.g. on macOS CurrentUser\My is the user's default keychain (generally login.keychain, but we ask for the default, not any keychain by name, and it can be replaced).

elinor-fung replied:
I guess current application (`AndroidKeyStore`) might actually be the 'natural' thing for Android apps, since they are all sandboxed and running under their own user ID.

elinor-fung replied:
Directed CurrentUser\My to use the store for the current application - currently just throwing `NotImplementedException` if it tries to add any certs with a private key.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.cs:
```diff
@@ -44,7 +44,47 @@ public static IExportPal LinkFromCertificateCollection(X509Certificate2Collectio
+            bool isReadWrite = (openFlags & OpenFlags.ReadWrite) == OpenFlags.ReadWrite;
+            if (isReadWrite && storeLocation == StoreLocation.LocalMachine)
+            {
+                // All LocalMachine stores are read-only
```
On Linux it's because we don't actually have a good way to update the stores (on some distros what we see is more or less authoritative, on some it's not).  Can a user create new system entries in the KeyStores? If so, they don't have to be read-only.

elinor-fung replied:
No, the system entries are entirely locked down from an app's perspective.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.cs:
```diff
@@ -44,7 +44,47 @@ public static IExportPal LinkFromCertificateCollection(X509Certificate2Collectio
+                    if (ordinalIgnoreCase.Equals(X509Store.DisallowedStoreName, storeName))
+                        return new UnsupportedDisallowedStore(openFlags);
+
+                    if (ordinalIgnoreCase.Equals(X509Store.RootStoreName, storeName))
+                    {
+                        if (isReadWrite)
+                            throw new CryptographicException(SR.Security_AccessDenied);
+
+                        return new TrustedStore(storeLocation);
+                    }
```
Can the application not add things to the AndroidCAStore?

elinor-fung replied:
It can add certs through Android's Intent/Activity mechanism by kicking off an activity to request the user add the cert:
<img width="174" alt="Screen Shot 2021-03-01 at 12 40 16 PM" src="https://user-images.githubusercontent.com/47805090/109560845-33a17d00-7a91-11eb-980e-25c6befdbfaf.png">

I haven't figured out how to properly deal with the async/callback nature of activities and their results through JNI though, so I have it read-only (at least for now).

elinor-fung replied:
Oh, actually, I just realised that this isn't possible in Android 11. They blocked being able to request this from a regular app - the user has to go through the phone's settings to manually install.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.cs:
```diff
@@ -44,7 +44,47 @@ public static IExportPal LinkFromCertificateCollection(X509Certificate2Collectio
+
+                        return new TrustedStore(storeLocation);
+                    }
+                    break;
```
Is `AndroidCAStore` just the roots, or does it also contain an intermediate CA cache?

elinor-fung replied:
My understanding is it is just the roots. I didn't find anything exposing an intermediate CA cache (seems to be an implementation detail of the default trust manager).

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509store.c:
```diff
@@ -0,0 +1,76 @@
+        if (!systemOnly || strncmp(aliasPtr, systemPrefix, sizeof(systemPrefix)/sizeof(char)) == 0)
+        {
+            jobject cert = (*env)->CallObjectMethod(env, loc[store], g_KeyStoreGetCertificate, alias);
+            assert(cert != NULL && !CheckJNIExceptions(env));
```
Maybe the assert makes sense in context, but it doesn't hold in general... if there is a key-only entry in the KeyStore then getCertificate() returns null.

It's probably better to just go ahead and accept that it can return null (and not fire the callback in that case).

elinor-fung replied:
It only makes sense for `AndroidCAStore`. I can switch it to handling null.

elinor-fung committed:
PR feedback

elinor-fung committed:
Use AndroidKeyStore for My/CurrentUser

Doesn't handle certs with private keys yet.

elinor-fung committed:
Update test

elinor-fung committed:
Handle private keys in CurrentUser/My store

elinor-fung committed:
Add tests

elinor-fung commented:
I've rebased this to the latest, updated it handle `CurrentUser\My` reasonably, and added tests (mostly re-used the OSX ones). Custom stores are not implemented - will throw PNSE about not being able to create the store.

I think this is ready for another review.

elinor-fung committed:
Fix release build

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509store.c:
```diff
@@ -0,0 +1,375 @@
+            return FAIL;
+    }
+
+    loc[alias] = CreateAliasForCertificate(env, cert);
```
Adding just the cert looks for a current alias, but with a private key doesn't?

elinor-fung replied:
This was relying on the store's behaviour of replacing existing entries (cert or private key) when setting a private key entry. That doesn't look to be the public contract of stores in general - I'll make it explicitly check like with just the cert (and remove/replace).

elinor-fung commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509store.c:
```diff
@@ -0,0 +1,375 @@
+    int hashCode = (*env)->CallIntMethod(env, cert, g_X509CertHashCode);
+
+    char buffer[9] = {0}; // 8-character hex + null terminator
+    size_t written = (size_t)snprintf(buffer, sizeof(buffer), "%08X", hashCode);
```
Will switch to sending the hash string as the alias from the managed side.

elinor-fung replied:
Passing down the hash string as the unique identifier / alias now. I didn't fully handle the case where the alias already exists and doesn't map to the expected cert - it currently errors if the existing entry doesn't match the cert, so the add will fail. It doesn't try to construct a new alias like what the Unix implementation does for file names. I could do something like return a different error code and keep trying to add with new aliases (and some corresponding logic for remove) - not sure how likely the collisions would be / important they should be to handle?

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509store.c:
```diff
@@ -0,0 +1,375 @@
+
+            PAL_KeyAlgorithm keyAlgorithm = PAL_UnknownAlgorithm;
+            void* keyHandle = HandleFromKeys(env, loc[publicKey], loc[privateKey], &keyAlgorithm);
+            if (keyHandle != NULL)
+            {
+                cb(AddGRef(env, loc[cert]), keyHandle, keyAlgorithm, context);
+            }
```
This feels like it should be

```
if (keyHandle != NULL && loc[cert] != NULL)
{
   current cb
}
else if (loc[cert] != NULL)
{
   cb with cert and no key
}
```

That is

* Presumably a key-only shouldn't fire the callback with a null cert
* An unrecognized key algorithm seems like it shouldn't stop the cert from being emitted.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.AndroidKeyStore.cs:
```diff
@@ -0,0 +1,149 @@
+            public static AndroidKeyStore OpenDefault(OpenFlags openFlags)
+            {
+                SafeX509StoreHandle store = Interop.AndroidCrypto.X509StoreOpenDefault();
+                if (store.IsInvalid)
+                    throw new CryptographicException();
```
```suggestion
                if (store.IsInvalid)
                {
                    store.Dispose();
                    throw new CryptographicException();
                }
```

No reason to send it off to the finalizer queue.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.AndroidKeyStore.cs:
```diff
@@ -0,0 +1,149 @@
+
+                bool success = Interop.AndroidCrypto.X509StoreRemoveCertificate(_keyStoreHandle, certPal.SafeHandle);
+                if (!success)
+                    throw new CryptographicException(SR.Cryptography_X509_StoreRemoveFailure);
```
I feel like Windows has cases where remove can look like it should have worked, but doesn't remove something, and no exception is thrown (which is why we don't already have a message).

That said, I'm fine with having an exception here, we can always remove it later.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/X509StoreMutableTests.Android.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace System.Security.Cryptography.X509Certificates.Tests
+{
+    [PlatformSpecific(TestPlatforms.Android)]
```
This doesn't seem to be needed, since it's in a conditionally included file.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/X509StoreMutableTests.Android.cs:
```diff
@@ -0,0 +1,79 @@
+                // Add
+                store.Add(cert);
+                Assert.True(IsCertInStore(cert, store), "Certificate should be found after add");
+                Assert.True(StoreHasPrivateKey(store, cert), "Certificate in store should have a private key");
```
Is there also a test that adds a cert as a plain cert, then adds it with a private key, to see that it correctly joins them?  (Perhaps it's in the other partial file already?)

elinor-fung replied:
I believe that is `AddPrivateAfterPublic` in the other partial file.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/X509StoreMutableTests.Android.cs:
```diff
@@ -0,0 +1,79 @@
+        public static bool PermissionsAllowStoreWrite { get; } = true;
+
+        [Fact]
+        public static void AddRemove_Cert()
```
This test looks just like VerifyRemove from the also-included partial file.  What's the difference?

elinor-fung replied:
I was going for one that just does a plain cert. I could collapse this into VerifyRemove and make that have a cert-only version? (Or remove entirely if the plain cert test is not interesting to separate out)

bartonjs replied:
Ah.  Let's move this concept into the shared file and do one of

* Make the existing test do both things
* Make the existing test a theory with a `bool withPrivateKey`
* Make two tests, VerifyRemove_CertWithKey, VerifyRemove_CertOnly (or similar naming)

elinor-fung committed:
Update src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.AndroidKeyStore.cs

Co-authored-by: Jeremy Barton <jbarton@microsoft.com>

elinor-fung committed:
Pass cert hash string from managed to use as alias in store

elinor-fung committed:
PR feedback
